core/shared/src/main/scala/zio/ZIOConcurrent.scala

```diff
@@ -0,0 +1,153 @@
+package zio
+
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.ConcurrentLinkedQueue
+import scala.annotation.tailrec
+import zio.Synchronizer._
+
+final class CountDownLatch(count: Int, maxWaiters: Option[Int]) {
+  private[this] val sync = Synchronizer(count, maxWaiters)
+
+  // TODO is it possible to evaluate once and reuse the Promise?
+  private[this] val awaitZero = sync.acquire(count)
+
+  def await: UIO[Unit] = awaitZero.asInstanceOf[UIO[Unit]]
+
+  // TODO stop at zero
+  def countDown(): UIO[Unit] = sync.release(1)
+}
+
+final class AsyncSemaphore(initialPermits: Int, maxWaiters: Option[Int]) {
+  private[this] val sync = Synchronizer(initialPermits, maxWaiters)
+
+  def apply[R, E, A](zio: ZIO[R, E, A]): ZIO[R, Either[E, Full], A] =
+    sync
+      .acquire(1)
+      .mapError(Right(_))
+      .flatMap { _ =>
+        zio
+          .mapError(Left(_))
+          .ensuring(sync.release(1))
+      }
+}
+
+final class ReentrantLock(maxWaiters: Option[Int]) {
+  @volatile private[this] var owner: Fiber.Id = null
+
+  private[this] val sem = new AsyncSemaphore(1, maxWaiters)
+
+  def apply[R, E <: Throwable, A](zio: ZIO[R, E, A]): ZIO[R, Either[E, Full], A] =
+    ZIO.effectSuspendTotalWith { (_, fiberId) =>
+      if (owner == fiberId) {
+        zio.mapError(Left(_))
+      } else {
+        sem(ZIO.effectSuspendTotal {
+          owner = fiberId
+          zio.ensuring(ZIO.effectTotal(owner = null))
+        })
+      }
+    }
+}
+
+final class Synchronizer(initialPermits: Int, maxWaiters: Int) {
+
+  // if >= 0, # of permits
+  // if < 0, # of waiters
+  private[this] val state = new AtomicInteger(initialPermits)
+
+  // TODO requires multiple allocations per waiter
+  private[this] val waiters = new ConcurrentLinkedQueue[(AtomicInteger, Promise[Full, Unit])]
+
+  def acquire(permits: Int): IO[Full, Unit] =
```

Comment:
If you like the idea of using `Ref`, your logic can stay the same, except for the loop that is taken care of by `Ref` itself. Something like this:
```scala
def acquire(permits: Permits): IO[Full, Unit] =
  IO.effectSuspendTotalWith { (_, fiberId) =>
    state.modify {
      case (available, waiters) if available >= permits =>
        IO.unit -> (available - permits, waiters)
      case (available, waiters) if waiters.size == maxWaiters =>
        Synchronizer.full -> (available, waiters)
      case (available, waiters) => 
        val p = Promise.unsafeMake[Nothing, Unit](fiberId)
        val rem = permits - available
        p.await -> (0, waiters.add((rem, p)))
    }.flatten
  }
```

`modify` is a method that allows you to atomically update the state and produce a result. Here the result is an IO action that must follow the update of the state. Technically this is not 100% atomic because the result of `modify` is `UIO[IO[Full, Unit]]` and so there is the flatmapping that needs to happen after that, so this can definitely be improved.